diff --git a/minecraft/net/minecraft/src/EntityBoat.java b/minecraft/net/minecraft/src/EntityBoat.java
index a7c529c..e5050f9 100644
--- a/minecraft/net/minecraft/src/EntityBoat.java
+++ b/minecraft/net/minecraft/src/EntityBoat.java
@@ -189,6 +189,41 @@ public class EntityBoat extends Entity
         this.velocityY = this.motionY = par3;
         this.velocityZ = this.motionZ = par5;
     }
+    
+    
+    public boolean IsPlayerHoldingSail(EntityPlayer player)
+    {
+        ItemStack var1 = player.inventory.mainInventory[player.inventory.currentItem];
+                int slot = player.inventory.currentItem;
+                ItemStack nextStack = player.inventory.getStackInSlot((slot+1)%9);
+                ItemStack prevStack = player.inventory.getStackInSlot((slot+8)%9);
+                boolean o = player.inventory.getCurrentItem() != null &&
+                    player.inventory.getCurrentItem().itemID == Item.map.itemID &&
+                    (
+                        (nextStack != null && nextStack.itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID) ||
+                        (prevStack != null && prevStack.itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID)
+                    );
+                return (var1 != null ? var1.itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID : false) || o;
+    }
+
+ // Copied from EntityPlayer.java
+    public boolean AppliesConstantForceWhenRidingBoat(EntityPlayer player)
+    {
+        return this.IsPlayerHoldingSail(player);
+    }
+
+    public double MovementModifierWhenRidingBoat(EntityPlayer player)
+    {
+        double var1 = 0.35D;
+
+        if (this.IsPlayerHoldingSail(player))
+        {
+            var1 = 1.0D;
+        }
+
+        return var1;
+    }
+    
 
     /**
      * Called to update the entity's position/logic.
@@ -319,11 +354,22 @@ public class EntityBoat extends Entity
 
             if (this.riddenByEntity != null)
             {
-                var14 *= this.riddenByEntity.MovementModifierWhenRidingBoat();
+                boolean constantForce = false;
+                if(this.riddenByEntity instanceof EntityPlayer)
+                {
+                    var14 *= MovementModifierWhenRidingBoat((EntityPlayer)this.riddenByEntity);
+                    constantForce = AppliesConstantForceWhenRidingBoat((EntityPlayer)this.riddenByEntity);
+                }
+                else
+                {
+                    var14 *= this.riddenByEntity.MovementModifierWhenRidingBoat();
+                    constantForce = this.riddenByEntity.AppliesConstantForceWhenRidingBoat();
+                }
+                
                 this.motionX += this.riddenByEntity.motionX * this.speedMultiplier;
                 this.motionZ += this.riddenByEntity.motionZ * this.speedMultiplier;
 
-                if (this.riddenByEntity.AppliesConstantForceWhenRidingBoat())
+                if (constantForce)
                 {
                     this.motionX -= Math.cos((double)this.rotationYaw * Math.PI / 180.0D) * var14 * 0.02D;
                     this.motionZ -= Math.sin((double)this.rotationYaw * Math.PI / 180.0D) * var14 * 0.02D;
diff --git a/minecraft/net/minecraft/src/ItemMap.java b/minecraft/net/minecraft/src/ItemMap.java
index 33db091..7e1d0b7 100644
--- a/minecraft/net/minecraft/src/ItemMap.java
+++ b/minecraft/net/minecraft/src/ItemMap.java
@@ -273,7 +273,19 @@ public class ItemMap extends ItemMapBase
                 var6.updateVisiblePlayers(var3, var1);
             }
 
-            if (var5)
+            InventoryPlayer inventory = var3.inventory;
+            int slot = inventory.currentItem;
+            ItemStack nextStack = inventory.getStackInSlot((slot+1)%9);
+            ItemStack prevStack = inventory.getStackInSlot((slot+8)%9);
+            boolean o =
+                var3.ridingEntity instanceof EntityBoat &&
+                inventory.getCurrentItem() != null &&
+                inventory.getCurrentItem().itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID &&
+                (
+                    (nextStack != null && this.itemID == nextStack.itemID) ||
+                    (prevStack != null && this.itemID == prevStack.itemID)
+                );
+            if (var5 || o)
             {
                 this.updateMapData(var2, var3, var6);
             }
diff --git a/minecraft_server/net/minecraft/src/EntityBoat.java b/minecraft_server/net/minecraft/src/EntityBoat.java
index 13fc362..ea5857a 100644
--- a/minecraft_server/net/minecraft/src/EntityBoat.java
+++ b/minecraft_server/net/minecraft/src/EntityBoat.java
@@ -132,6 +132,39 @@ public class EntityBoat extends Entity
         return !this.isDead;
     }
 
+    public boolean IsPlayerHoldingSail(EntityPlayer player)
+    {
+        ItemStack var1 = player.inventory.mainInventory[player.inventory.currentItem];
+                int slot = player.inventory.currentItem;
+                ItemStack nextStack = player.inventory.getStackInSlot((slot+1)%9);
+                ItemStack prevStack = player.inventory.getStackInSlot((slot+8)%9);
+                boolean o = player.inventory.getCurrentItem() != null &&
+                    player.inventory.getCurrentItem().itemID == Item.map.itemID &&
+                    (
+                        (nextStack != null && nextStack.itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID) ||
+                        (prevStack != null && prevStack.itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID)
+                    );
+                return (var1 != null ? var1.itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID : false) || o;
+    }
+
+ // Copied from EntityPlayer.java
+    public boolean AppliesConstantForceWhenRidingBoat(EntityPlayer player)
+    {
+        return this.IsPlayerHoldingSail(player);
+    }
+
+    public double MovementModifierWhenRidingBoat(EntityPlayer player)
+    {
+        double var1 = 0.35D;
+
+        if (this.IsPlayerHoldingSail(player))
+        {
+            var1 = 1.0D;
+        }
+
+        return var1;
+    }
+    
     /**
      * Called to update the entity's position/logic.
      */
@@ -261,11 +294,22 @@ public class EntityBoat extends Entity
 
             if (this.riddenByEntity != null)
             {
-                var14 *= this.riddenByEntity.MovementModifierWhenRidingBoat();
+                boolean constantForce = false;
+                if(this.riddenByEntity instanceof EntityPlayer)
+                {
+                    var14 *= MovementModifierWhenRidingBoat((EntityPlayer)this.riddenByEntity);
+                    constantForce = AppliesConstantForceWhenRidingBoat((EntityPlayer)this.riddenByEntity);
+                }
+                else
+                {
+                    var14 *= this.riddenByEntity.MovementModifierWhenRidingBoat();
+                    constantForce = this.riddenByEntity.AppliesConstantForceWhenRidingBoat();
+                }
+                
                 this.motionX += this.riddenByEntity.motionX * this.speedMultiplier;
                 this.motionZ += this.riddenByEntity.motionZ * this.speedMultiplier;
 
-                if (this.riddenByEntity.AppliesConstantForceWhenRidingBoat())
+                if (constantForce)
                 {
                     this.motionX -= Math.cos((double)this.rotationYaw * Math.PI / 180.0D) * var14 * 0.02D;
                     this.motionZ -= Math.sin((double)this.rotationYaw * Math.PI / 180.0D) * var14 * 0.02D;
diff --git a/minecraft_server/net/minecraft/src/ItemMap.java b/minecraft_server/net/minecraft/src/ItemMap.java
index dcc2220..c1b6d62 100644
--- a/minecraft_server/net/minecraft/src/ItemMap.java
+++ b/minecraft_server/net/minecraft/src/ItemMap.java
@@ -257,7 +257,19 @@ public class ItemMap extends ItemMapBase
                 var6.updateVisiblePlayers(var3, var1);
             }
 
-            if (var5)
+            InventoryPlayer inventory = var3.inventory;
+            int slot = inventory.currentItem;
+            ItemStack nextStack = inventory.getStackInSlot((slot+1)%9);
+            ItemStack prevStack = inventory.getStackInSlot((slot+8)%9);
+            boolean o =
+                var3.ridingEntity instanceof EntityBoat &&
+                inventory.getCurrentItem() != null &&
+                inventory.getCurrentItem().itemID == FCBetterThanWolves.fcItemWindMillBlade.itemID &&
+                (
+                    (nextStack != null && this.itemID == nextStack.itemID) ||
+                    (prevStack != null && this.itemID == prevStack.itemID)
+                );
+            if (var5 || o)
             {
                 this.updateMapData(var2, var3, var6);
             }
